lightning/src/offers/invoice_request.rs

```diff
@@ -9,18 +9,121 @@
-use bitcoin::secp256k1::PublicKey;
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::{Message, PublicKey, self};
 use bitcoin::secp256k1::schnorr::Signature;
 use core::convert::TryFrom;
 use core::str::FromStr;
 use io;
 use ln::features::OfferFeatures;
 use offers::merkle::{SignatureTlvStream, self};
-use offers::offer::{Amount, OfferContents, OfferTlvStream, self};
+use offers::offer::{Offer, OfferContents, OfferTlvStream, self};
 use offers::parse::{Bech32Encode, ParseError, SemanticError};
 use offers::payer::{PayerContents, PayerTlvStream, self};
-use util::ser::{Writeable, Writer};
+use util::ser::{WithoutLength, Writeable, Writer};
+
+///
+const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "signature");
+
+///
+pub struct InvoiceRequestBuilder<'a> {
+	offer: &'a Offer,
+	invoice_request: InvoiceRequestContents,
+}
+
+impl<'a> InvoiceRequestBuilder<'a> {
+	pub(super) fn new(offer: &'a Offer, payer_id: PublicKey) -> Self {
+		Self {
+			offer,
+			invoice_request: InvoiceRequestContents {
+				payer: PayerContents(None), offer: offer.contents.clone(), chain: None,
+				amount_msats: None, features: None, quantity: None, payer_id, payer_note: None,
+				signature: None,
+			},
+		}
+	}
+
+	///
+	pub fn payer_info(mut self, payer_info: Vec<u8>) -> Self {
+		self.invoice_request.payer = PayerContents(Some(payer_info));
+		self
+	}
+
+	///
+	pub fn chain(mut self, network: Network) -> Result<Self, SemanticError> {
+		let block_hash = genesis_block(network).block_hash();
+		if !self.offer.supports_chain(block_hash) {
+			return Err(SemanticError::UnsupportedChain)
+		}
+
+		self.invoice_request.chain = Some(block_hash);
+		Ok(self)
+	}
+
+	///
+	pub fn amount_msats(mut self, amount_msats: u64) -> Result<Self, SemanticError> {
+		if !self.offer.is_sufficient_amount(amount_msats) {
+			return Err(SemanticError::InsufficientAmount);
+		}
+
+		self.invoice_request.amount_msats = Some(amount_msats);
+		Ok(self)
+	}
+
+	///
+	pub fn features(mut self, features: OfferFeatures) -> Self {
+		self.invoice_request.features = Some(features);
+		self
+	}
+
+	///
+	pub fn quantity(mut self, quantity: u64) -> Result<Self, SemanticError> {
+		if !self.offer.is_valid_quantity(quantity) {
+			return Err(SemanticError::InvalidQuantity);
+		}
+
+		self.invoice_request.quantity = Some(quantity);
+		Ok(self)
+	}
+
+	///
+	pub fn payer_note(mut self, payer_note: String) -> Self {
+		self.invoice_request.payer_note = Some(payer_note);
+		self
+	}
+
+	///
+	pub fn build_signed<F>(mut self, sign: F) -> Result<InvoiceRequest, secp256k1::Error>
+	where F: FnOnce(&Message) -> Signature
+	{
+		// Use the offer bytes instead of the offer TLV stream as the offer may have contained
+		// unknown TLV records, which are not stored in `OfferContents`.
+		let (payer_tlv_stream, _offer_tlv_stream, invoice_request_tlv_stream, _) =
+			self.invoice_request.as_tlv_stream();
+		let offer_bytes = WithoutLength(&self.offer.bytes);
```

Comment:
Note, this may contain custom records after the signature range. So actually, need to break this into two sequences of bytes. When building an `invoice` from an `invoice_request` will need to three sequences 0-160 (`offer` and `invoice_request`), 240-1000 (signatures to strip), and 1000+ (custom records).

Reply:
>  and 1000+ (custom records).

Hm I can't find this in the spec, am I missing it? Commented on the PR: https://github.com/rustyrussell/lightning-rfc/pull/7#discussion_r961244111

Reply:
Ah, right. As written it does not support it. I suppose unused types in the 0-240 range could be custom, but presumably we may want a dedicated range. BTW, that was suppose to read 0-240 in my earlier comment.